src/Agent.Listener/SelfUpdater.cs

```diff
@@ -155,6 +166,47 @@ private async Task<bool> UpdateNeeded(string targetVersion, CancellationToken to
             return false;
         }
 
+        private async Task<bool> HashValidation(string archiveFile)
+        {
+            if (_hashValidationDisabled)
+            {
+                Trace.Info($"Agent package hash validation disabled, so skipping it");
+                return true;
+            }
+
+            // Determine the service deployment type based on connection data. (Hosted/OnPremises)
+            var connectionData = await ConfigurationManager.GetConnectionData(_serverUrl, _creds, _locationServer);
```

Comment:
Could we use generic vss connection from AgentServer instead?
https://github.com/microsoft/azure-pipelines-agent/blob/master/src/Microsoft.VisualStudio.Services.Agent/AgentServer.cs#L62
ConfigurationManager is mostly for agent configuration flow - I would separate it from the execution flow.

Reply:
I would also put IsHosted logic to some separate third module - since this would be some sort of generic check.

Reply:
I moved the `IsHosted` method logic to [a separate module](https://github.com/microsoft/azure-pipelines-agent/blob/67963581c0c8dad95d6aaa15dc34c143287da465/src/Microsoft.VisualStudio.Services.Agent/Util/ServerUtil.cs). Since we cannot reuse the `_genericConnection` from the `AgentServer` here, we have to create a new one, but only once, when we determine the deployment type the first time.